File: Java-SpringBoot/BE-01-InMemory/java-21-springboot-4.0/src/main/java/com/flyrank/userservice/repository/InMemoryUserRepository.java
package com.flyrank.userservice.repository;

import com.flyrank.userservice.dto.UserResponse;
import org.springframework.stereotype.Repository;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Thread-safe In-Memory UserRepository implementation using ConcurrentHashMap and LinkedHashMap for SequencedCollection support.
 */
@Repository
public class InMemoryUserRepository implements UserRepository {

    // ConcurrentHashMap for fast thread-safe UUID lookup
    private final Map<UUID, UserResponse> userStore = new ConcurrentHashMap<>();

    @Override
    public UserResponse save(UserResponse user) {
        userStore.put(user.id(), user);
        return user;
    }

    @Override
    public Optional<UserResponse> findById(UUID id) {
        return Optional.ofNullable(userStore.get(id));
    }

    @Override
    public SequencedCollection<UserResponse> findAllSequenced() {
        // Java 21 SequencedCollection interface
        return new ArrayList<>(userStore.values());
    }
}

File: Java-SpringBoot/BE-01-InMemory/kotlin-2.3-springboot-3.5/src/main/kotlin/com/flyrank/userservice/repository/InMemoryUserRepository.kt
package com.flyrank.userservice.repository

import com.flyrank.userservice.dto.UserResponse
import org.springframework.stereotype.Repository
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Thread-safe In-Memory UserRepository implementation in Kotlin using ConcurrentHashMap.
 */
@Repository
class InMemoryUserRepository : UserRepository {

    private val userStore = ConcurrentHashMap<UUID, UserResponse>()

    override fun save(user: UserResponse): UserResponse {
        userStore[user.id] = user
        return user
    }

    override fun findById(id: UUID): UserResponse? {
        return userStore[id]
    }

    override fun findAll(): List<UserResponse> {
        return userStore.values.toList()
    }
}
